### ISSUE TITLE
WinHttpHandler causing “Unable to load DLL 'api-ms-win-core-localization-l1-2-0.dll'” error on production server

### ISSUE THREAD

**Issue by markvincze** [dotnet/corefx#8091]
We've started using the [WinHttpHandler](https://www.nuget.org/packages/System.Net.Http.WinHttpHandler/) NuGet package in an ASP.NET application. On the development machines it's working fine, and also works on our staging environment. The application targets .NET 4.6.1.

However, on production it throws the following error:

    System.DllNotFoundException: Unable to load DLL 'api-ms-win-core-localization-l1-2-0.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)

I have googled quite a lot about this problem, but couldn't find any meaningful information about this dll. (I found in forums that some people manually copied the dlls to their production servers when having similar issues, but that doesn't seem to be the proper solution.)

Where is this dll coming from? Should it be explicitly installed?

(On my dev machine I found it in `C:\Windows\System32`, on our staging machine it's in `C:\Windows\System32\downlevel`, but I don't know where it's coming from.)

**Comment by joshfree** [comment]
cc: @chcosta @ericstj

**Comment by ericstj** [comment]
The WinHttpHandler uses API sets even in the desktop implementation. @davidsh you might want to consider cross-compiling for desktop so that you don't require API-sets to be installed.

For .NET Core we ship the API sets out of band as part of the runtime in package Microsoft.NETCore.Windows.ApiSets.  The package requires the app to be built arch specific with project.json declaring the minimum target platform in the runtime section of project.json (eg: win7-x86, win7-x64).

@weshaggard what is our plan for corefx libraries on desktop using API sets?  Do we want folks to a) cross compile, b) depend on the api-sets package, c) document some OS-update as a pre-requisite.

**Comment by davidsh** [comment]
>The WinHttpHandler uses API sets even in the desktop implementation. @davidsh you might want to consider cross-compiling for desktop so that you don't require API-sets to be installed.

We might want/need to do something similar for System.Net.Http package which uses WinHttpHandler internally.

**Comment by davidsh** [comment]
cc: @CIPop @himadrisarkar

**Comment by ericstj** [comment]
I'd like to hear from @weshaggard and @gkhanna79 what the overall strategy is for PInvokes on Windows for portable libraries.  Are we still getting any value out of API sets?

**Comment by markvincze** [comment]
Thanks for quickly responding.  
Just to make it clear: this is not a .NET Core application, but an ASP.NET 4.6 app, targeting .NET 4.6.1.  
If it matters, the OS on the production system where the problem occurs is Windows Server 2008 R2.

**Comment by ericstj** [comment]
@markvincze understood.  The problem here is that all the CoreFx assemblies PInvoke into API set dlls rather than the actual Windows dlls.  Those API set DLLs are not part of the operating system on older OSes.  For .NET Core we redistribute the API sets, but that is not done for .NET desktop.

**Comment by gkhanna79** [comment]
For RC2, .NET Core installation carries api-set binaries. For RTM, we are working towards a plan where they will not be required and link against standard import libraries (e.g. kernel32.lib). However, @weshaggard is aware of certain libraries that need to run on platforms where api-set linkage is required and such libraries would need to be built differently.

**Comment by markvincze** [comment]
Is there any way to solve this on the short term and use this package on Windows Server 2008 R2?
It is urgent for me to get this working, so I'm interested in temporary or short-term solutions too.

For example it's working on my development machine and I found the dlls in `C:\Windows\System32`, yet my machine has Windows 7 installed.
How did those dlls get installed on my computer? Can I also install them to Windows Server 2008 R2?

**Comment by davidsh** [comment]
api-set binaries are something that was added in the Windows 8 (and Windows Server 2012) timeframe. These binaries don't exist on platforms such as Windows 7 (or Windows Server 2008 R2).  

That is why the various .NET Core installers will copy those binaries to the machine.

>How did those dlls get installed on my computer? Can I also install them to Windows Server 2008 R2?

Yes, you can install them.  But you should do it as part of installing .NET Core tools.  I realize that you have a Desktop application and not ASP.NET app, but that should be ok.  You could probably then manually copy those set of binaries to your folder where your Desktop EXE is. Then it will be able to find the api-set binaries.  They do not need to be in `windows\system32` directory. They can be app-local to your executable.

Longer term we are working on a better solution for using these NuGet pacakges on Desktop to handle Windows 7 (and Windows 2008 R2) platforms.

**Comment by ericstj** [comment]
@markvincze are you using project.json or packages.config to build your app?

**Comment by ericstj** [comment]
> How did those dlls get installed on my computer? Can I also install them to Windows Server 2008 R2

Some of them are installed by the UCRT update: https://support.microsoft.com/en-us/kb/2999226
Not all though.  I don't think this has LibraryLoader which is required by WinHttpHandler.  There appear to be some other updates like https://support.microsoft.com/en-us/kb/2790113 which include this, but I'm not sure of a comprehensive one.

If you're using project.json you can just reference the Microsoft.NETCore.Windows.ApiSets package from your app and be sure to deploy for runtime win7-x86 or win7-x64, as appropriate.

If you're using packages.config you'll need to manually download and extract the packages, as appropriate for the bitness of your app:
[runtime.win7-x64.Microsoft.NETCore.Windows.ApiSets](https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.win7-x64.Microsoft.NETCore.Windows.ApiSets/1.0.1-rc3-24026-00)
[runtime.win7-x86.Microsoft.NETCore.Windows.ApiSets](https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.win7-x86.Microsoft.NETCore.Windows.ApiSets/1.0.1-rc3-24026-00)

**Comment by markvincze** [comment]
It's an ASP.NET 4.6 app, and I'm using csproj and packages.config, and not the new project.json model.

I'll experiment on a clean Windows Server 2008 R2 virtual machine and I'll write a comment about what approach worked.

**Comment by weshaggard** [comment]
>I'd like to hear from @weshaggard and @gkhanna79 what the overall strategy is for PInvokes on Windows for portable libraries. Are we still getting any value out of API sets?

Unfortunately I don't think we really have any good strategies for portable libraries that do PInvoke. For the cases like the networking library I assume we are already cross-compiling and if we are then we should probably just not use api-sets for that build.

**Comment by markvincze** [comment]
I don't know how long term this solution is, but installing [Visual C++ Redistributable for Visual Studio 2015](https://chocolatey.org/packages/vcredist2015) solved the issue.

(This is what I installed:

1. Clean Windows Server 2008 R2
2. All Windows updates
3. Windows Server 2008 R2 Service Pack 1
4. Web server role (to get IIS) with the Application Development features
5. .NET Framework Redistributable 4.6.1  
At this point my site was working but I got the `Unable to load DLL 'api-ms-win-core-localization-l1-2-0.dll'` error when using `WinHttpHandler`.
6. Visual C++ Redistributable for Visual Studio 2015  
Now `WinHttpHandler` is working properly.)

**Comment by davidsh** [comment]
@weshaggard 
>For the cases like the networking library I assume we are already cross-compiling and if we are then we should probably just not use api-sets for that build. 

I don't know what you mean by "cross-compiling". In terms of WinHttpHandler CSPROJ and source files, we simply use the Common/src/ Interop source files with have p/invoke definitions that reference various dlls including api-set dlls. We would have to add `ifdef  NET46` type of lines to those source files and pick different dll's to reference for the p/invokes

So, it is more than just "cross-compiling".

**Comment by ericstj** [comment]
@davidsh You don't need to ifdef.  Just add a net46 configuration and swap out all the Common/src/ Interop files with ones that don't use the api-sets.

**Comment by davidsh** [comment]
@ericstj 
>You don't need to ifdef. Just add a net46 configuration and swap out all the Common/src/ Interop files with ones that don't use the api-sets.

Yes, you're right about that I don't have to use ifdef. But using different Interop files (which don't exist yet) is basically the same thing. :)

The bottom line is how important is this? Especially given a solution like this:
> installing Visual C++ Redistributable for Visual Studio 2015 solved the issue.

That might be the easiest/best solution for customers long term.

**Comment by ericstj** [comment]
@markvincze yeah the VC redist contains that UCRT OS update I was pointing at before.  It doesn't actually contain api-ms-win-core-libraryloader-l1-1-0.dll which I was pointing out before.  Looks like WinHttpHandler hits that in this error codepath: https://github.com/dotnet/corefx/blob/0f0a4bcc03db4bc709e100b2107d241f826dd1bb/src/Common/src/System/Net/Http/WinHttpException.cs#L58
So you might have just gotten lucky and not hit that path in your service.

**Comment by ericstj** [comment]
@davidsh the VC redist is far from complete.  It only contains a small subset of the API set DLLs.  You can see the list in the KB article I linked.  I'm not aware of any OS update which has a comprehensive list of API set DLLs.

If we do what @gkhanna79 suggests and
>  link against standard import libraries

We won't have to worry about this problem at all. In that case we'd be changing all the common interop source back to the standard implementation libraries.  @gkhanna79 @weshaggard is that the plan?

**Comment by davidsh** [comment]
One other thing to consider. While we could have a NET46 version that doesn't use api-sets, that is not recommended really.  While it would result in everything working fine on older Windows platforms like Windows 7/Windows Server 2008 R2, it might stop working in future Windows versions.  Windows created api-sets to allow operating system binary refactoring.  And current guidelines are that api-sets are the best-practice.  Since there is not any way to target "old" Windows vs. "new" Windows operating system versions w.r.t. NET46, moving away from using api-sets is not the best solutiion.

cc: @stephentoub

**Comment by davidsh** [comment]
@ericstj Based on latest plan, library code will remain as-is and p/invoke to api-sets.  That means that this particular bug isn't actionable via changing any WinHttpHandler code/package. So, assigning to you for now since this is a tooling/sdk install issue.

**Comment by joshfree** [comment]
At a bare minimum we need to document this for 1.0
cc: @leecow

**Comment by joshfree** [comment]
@gkhanna79 can you summarize exactly how customers will need to install these API-sets and find the install links please?

**Comment by gkhanna79** [comment]
Assigning to @weshaggard who has been driving guidance for FX/Library authors.

**Comment by weshaggard** [comment]
The below will go into the release notes.

**Issue:**
Some libraries that Pinvoke into api-set's and target .NET Framework in our nuget packages might fail to run on Windows 7. 

**Workarounds:**
Some of the api-sets are installed by the UCRT update: https://support.microsoft.com/en-us/kb/2999226
Not all though. I don't think this has LibraryLoader which is required by WinHttpHandler. There appear to be some other updates like https://support.microsoft.com/en-us/kb/2790113  which include this, but I'm not sure of a comprehensive one.

If you're using project.json you can just reference the Microsoft.NETCore.Windows.ApiSets package from your app and be sure to deploy for runtime win7-x86 or win7-x64, as appropriate.

If you're using packages.config you'll need to manually download and extract the packages, as appropriate for the bitness of your app:
[runtime.win7-x64.Microsoft.NETCore.Windows.ApiSets](https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.win7-x64.Microsoft.NETCore.Windows.ApiSets/1.0.1)
[runtime.win7-x86.Microsoft.NETCore.Windows.ApiSets](https://dotnet.myget.org/F/dotnet-core/api/v2/package/runtime.win7-x86.Microsoft.NETCore.Windows.ApiSets/1.0.1)

**Comment by weshaggard** [comment]
cc @joperezr

**Comment by wiwichu** [comment]
I am trying to resolve this issue deploying a standalone app to Server 2008. Do you have an example of the project.json file with this correction?

Thank you

**Comment by markvincze** [comment]
Hey @wiwichu,

I didn't have to do any change on the project (actually, in my case it's not a .NET Core app using project.json, but it's a full .NET app using a csproj). The only thing I had to do is to install https://chocolatey.org/packages/vcredist2015 on the server.  
Doesn't that fix the issue for you?

**Comment by wiwichu** [comment]
markvincze,

I'm trying to create a standalone app which should allow me to deploy it with all the required binaries, so I'd like to get it working without making any changes to the server if I can. Also, I'm targeting NetCoreApp rather than .Net Framework, so it does not seem to be limited to just targeting .Net Framework as weshaggard mentioned in the release notes.

I added the Microsoft.NETCore.Windows.ApiSets package, and added win7-x86 under runtimes. However when I publish with that configuration, it doesn't find the folder netcoreapp1.0\win7-x86. Only win10-x86 is there. How does the win7-x86 folder get created?

**Comment by wiwichu** [comment]
I should add that I am using a native binary from a C++ project. Could this be the piece which requires api-ms-win-core-registry-l1-1-0.dl? I didn't think so since it happens right at startup. I'm pasting the complete exception details below:

Unhandled Exception: System.Reflection.TargetInvocationException: Exception has

been thrown by the target of an invocation. ---> System.DllNotFoundException: Un

able to load DLL 'api-ms-win-core-registry-l1-1-0.dll': The specified module cou

ld not be found. (Exception from HRESULT: 0x8007007E)

at Interop.mincore.RegOpenKeyEx(SafeRegistryHandle hKey, String lpSubKey, Int

32 ulOptions, Int32 samDesired, SafeRegistryHandle& hkResult)

   at Microsoft.Win32.RegistryKey.InternalOpenSubKey(String name, Int32 rights)

   at Microsoft.AspNetCore.DataProtection.RegistryPolicyResolver.ResolveDefaultP

olicy()

   at Microsoft.Extensions.DependencyInjection.DataProtectionServices.<GetDefaul

tServices>d__0.MoveNext()

   at Microsoft.Extensions.DependencyInjection.Extensions.ServiceCollectionDescr

iptorExtensions.TryAdd(IServiceCollection collection, IEnumerable`1 descriptors)



   at Microsoft.Extensions.DependencyInjection.DataProtectionServiceCollectionEx

tensions.AddDataProtection(IServiceCollection services)

   at Microsoft.Extensions.DependencyInjection.MvcViewFeaturesMvcCoreBuilderExte

nsions.AddViewServices(IServiceCollection services)

   at Microsoft.Extensions.DependencyInjection.MvcViewFeaturesMvcCoreBuilderExte

nsions.AddViews(IMvcCoreBuilder builder)

   at Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.Ad

dMvc(IServiceCollection services)

   at FinSysCore.Startup.ConfigureServices(IServiceCollection services)

   --- End of inner exception stack trace ---

   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments,

 Signature sig, Boolean constructor)

   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Objec

t[] parameters, Object[] arguments)

   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invoke

Attr, Binder binder, Object[] parameters, CultureInfo culture)

   at Microsoft.AspNetCore.Hosting.Internal.ConfigureServicesBuilder.Invoke(Obje

ct instance, IServiceCollection exportServices)

   at Microsoft.AspNetCore.Hosting.Internal.ConfigureServicesBuilder.<>c__Displa

yClass4_0.<Build>b__0(IServiceCollection services)

   at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.ConfigureServices(ISer

viceCollection services)

   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()

   at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()

   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()

   at FinSysCore.Program.Main(String[] args)

**Comment by markvincze** [comment]
@wiwichu,

I think this is a separate issue then. Can you show your `project.json`?  
I suspect that you need to add the `runtimes` section to your `project.json`. This is how mine looks in a service in which I'm bundling everything:

```
  "runtimes":{
    "win7-x64": { },
    "win8-x64": { },
    "win10-x64": { },

    "osx.10.10-x64": { },
    "osx.10.11-x64": { },

    "ubuntu.14.04-x64": { },
    "ubuntu.14.10-x64": { },
    "ubuntu.15.04-x64": { },
    "ubuntu.16.04-x64": { },
    "centos.7-x64": { },
    "debian.8-x64": { }
  },
``` 

Probably you need `"win7-x86": { },`.

Cheers,
Mark

**Comment by wiwichu** [comment]
My runtimes is like this:

  "runtimes": {
    "win10-x86": {},
    "win7-x86": {}
  },

I'll past the entire project.json below

The error I get when publishing is:

Could not find a part of the path 'C:\Users\Patrick\Documents\Visual Studio 2015\Projects\FinSys\FinSysCore\bin\Release\netcoreapp1.0\win7-x86\FinSysCore.dll'.

And indeed there is no win7-x86 folder, just win10-x86, as in the attached screenshot.

**Comment by wiwichu** [comment]
Here is the project.json:

{
  "buildOptions": {
    "emitEntryPoint": true,
    "preserveCompilationContext": true
  },
  "dependencies": {
    "AutoMapper": "4.2.1",
    "FinSysCoreCalc": "1.0.3",
    "Microsoft.NETCore.Windows.ApiSets": "1.0.1",
    "Microsoft.AspNetCore.Diagnostics": "1.0.0",
    "Microsoft.AspNetCore.Mvc": "1.0.0",
    "Microsoft.AspNetCore.Mvc.TagHelpers": "1.0.0",
    "Microsoft.AspNetCore.Mvc.ViewFeatures": "1.0.0",
    "Microsoft.AspNetCore.Server.IISIntegration": "1.0.0",
    "Microsoft.AspNetCore.Server.Kestrel": "1.0.0",
    "Microsoft.AspNetCore.StaticFiles": "1.0.0",
    "Microsoft.EntityFrameworkCore": "1.0.0",
    "Microsoft.EntityFrameworkCore.SqlServer": "1.0.0",
    "Microsoft.EntityFrameworkCore.Tools": {
      "version": "1.0.0-preview2-final",
      "type": "build"
    },
    "Microsoft.Extensions.Configuration.FileExtensions": "1.0.0",
    "Microsoft.Extensions.Configuration.Json": "1.0.0",
    "Microsoft.Extensions.Logging.Console": "1.0.0",
    "Microsoft.Extensions.Logging.Debug": "1.0.0"
  },
  "frameworks": {
    "netcoreapp1.0": {
      "dependencies": {
        "Microsoft.NETCore.App": {
          "version": "1.0.0"
        }
      },
      "imports": [
        "dotnet5.6",
        "portable-net45+win8"
      ]
    },
    "net461": {}
  },
  "runtimes": {
    "win10-x86": {},
    "win7-x86": {}
  },
  "publishOptions": {
    "include": [
      "wwwroot",
      "web.config",
      "Views",
      "config.json"
    ]
  },
  "runtimeOptions": {
    "configProperties": {
      "System.GC.Server": true
    }
  },
  "scripts": {
    "postpublish": [ "dotnet publish-iis --publish-folder %publish:OutputPath% --framework %publish:FullTargetFramework%" ]
  },
  "tools": {
    "Microsoft.AspNetCore.Server.IISIntegration.Tools": "1.0.0-preview2-final",
    "Microsoft.EntityFrameworkCore.Tools": {
      "version": "1.0.0-preview2-final",
      "imports": [
        "portable-net45+win8+dnxcore50",
        "portable-net45+win8"
      ]
    }
  }
}

**Comment by wiwichu** [comment]
And screenshot

![image](https://cloud.githubusercontent.com/assets/6375518/17780976/5599abc8-653b-11e6-8733-668d4e348dd2.png)

**Comment by markvincze** [comment]
@wiwichu,  
Can you try to `build` or `publish` with the flag `--runtime  win7-x86`?  
That should produce the output for the specific runtime.

**Comment by wiwichu** [comment]
I've been trying to publish and build from VS.

If I build from the command line like this:

dotnet publish --runtime win7-x86

Then it puts the win7-x86 folder debug. If I move it to release, it then seems to publish ok. It would be nice to be able to do it all from VS though.

When I deploy to the server, it gets past the point where it didn't find the apisets and starts ok, but then I get an error that my pinvoke does not find my native c++ dll which is in the same folder as the exe with the main method (works ok on my Win 10 workstation). I am suspecting another compatibility issue. I built it in VC targeting Win32.

**Comment by markvincze** [comment]
In order to get a release build you also need the flag `--configuration Release`.
I don't know what can cause the other problem. Can you share more information, like what kind of dll that is, and show the code with which you're trying to call into it?

**Comment by wiwichu** [comment]
I actually got it working. The other problem was I had to build a static cpp dll all so the dependent binaries are built in. Thanks for all the help. I may need more as I work on it.
